File: 3227-find-missing-and-repeated-values/find-missing-and-repeated-values.java
class Solution {
    public int[] findMissingAndRepeatedValues(int[][] grid) {
        HashSet <Integer> set = new HashSet <> ();
        int n = grid.length;
        int actualSum = 0;
        int expectedSum = (n*n *(n*n + 1)) / 2;
        int a = -1, b = -1;
        for (int i = 0; i<n; i++){
            for (int j = 0; j<n; j++){
                actualSum += grid[i][j];
                int val = grid[i][j];
                if (set.contains(val)){
                    a = val;
                }
                set.add(val);
            }
        }
        b = expectedSum - actualSum + a;
        return new int[]{a,b};
    }
}

File: 3227-find-missing-and-repeated-values/find-missing-and-repeated-values.cpp
class Solution {
public:
    vector<int> findMissingAndRepeatedValues(vector<vector<int>>& grid) {
        // unordered_set <int> seen;
        // vector <int> ans;

        // int actualSum = 0;
        // int n = grid.size();
        // int total = n * n;
        // long long expectedSum = 1LL * total * (total + 1) / 2;

        // for(int i = 0; i < n; i++){
        //     for(int j = 0; j < n; j++){
        //         if (seen.count(grid[i][j])) ans.push_back(grid[i][j]);
        //         else seen.insert(grid[i][j]);
        //         actualSum += grid[i][j];
        //     }
        // }

        // ans.push_back(expectedSum + ans[0] - actualSum);
        // return ans;

        int n = grid.size();
        int total = n * n;

        int actualSum = 0;
        long long expectedSum = 1LL * total * (total + 1)/2;

        long actualSqSum = 0;
        long long expectedSqSum = 1LL * total * (total + 1) * (2*total + 1) / 6;

        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                actualSum += grid[i][j];
                actualSqSum += grid[i][j] * grid[i][j];
            }
        }

        int diff = actualSum - expectedSum;
        int sum = (actualSqSum - expectedSqSum) / diff;

        int x = (diff + sum) / 2;
        int y = sum - x;

        return {x,y}; 
    }
};